csharp/src/lesson03/README.md

```diff
@@ -0,0 +1,327 @@
+var span = scope.Span
+    .SetTag("span.kind", "client")  // or: .SetTag(Tags.SpanKind, Tags.SpanKindClient);
+    .SetTag("http.method", "GET")   // or: .SetTag(Tags.HttpMethod, "GET");
+    .SetTag("http.url", url);       // or: .SetTag(Tags.HttpUrl, url);
+
+var dictionary = new Dictionary<string, string>();
+_tracer.Inject(span.Context, BuiltinFormats.HttpHeaders, new TextMapInjectAdapter(dictionary));
+foreach (var entry in dictionary)
+    _webClient.Headers.Add(entry.Key, entry.Value);
+```
+
+Notice that we also add a couple additional tags to the span with some metadata about the HTTP request,
+and we mark the span with a `span.kind=client` tag, as recommended by the OpenTracing
+[Semantic Conventions][semantic-conventions]. There are other tags we could add. All of the recommended tags 
+are specified in `OpenTracing.Tag.Tags` for easier usage as shown above.
+
+### Instrumenting the Server manually
+
+Our server is currently not instrumented for tracing. We need to do the following:
+
+#### Create an instance of a Tracer, similar to how we did it in `Hello.cs`
+
+Add a tracer property to the `Startup` class:
+
+```csharp
+private static readonly Tracer Tracer = Tracing.Init("Webservice");
+```
+
+Register the tracer and add it to the services, making it available globally through Dependency Injection:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddMvc();
+
+    services.AddSingleton<ITracer, Tracer>(t => Tracer);
+}
+```
+
+Add a member variable and a constructor to the `FormatController`. The `ITracer` will be supplied using 
+Dependency Injection:
+
+```csharp
+private readonly ITracer _tracer;
+
+public FormatController(ITracer tracer)
+{
+    _tracer = tracer;
+}
+```
+
+#### Extract the span context from the incoming request using `ITracer.Extract`
+
+First, add a helper function:
+
+```csharp
+public static IScope StartServerSpan(ITracer tracer, IDictionary<string, string> headers, string operationName)
+{
+    ISpanBuilder spanBuilder;
+    try
+    {
+        ISpanContext parentSpanCtx = tracer.Extract(BuiltinFormats.HttpHeaders, new TextMapExtractAdapter(headers));
+
+        spanBuilder = tracer.BuildSpan(operationName);
+        if (parentSpanCtx != null)
+        {
+            spanBuilder = spanBuilder.AsChildOf(parentSpanCtx);
+        }
+    }
+    catch (Exception)
+    {
+        spanBuilder = tracer.BuildSpan(operationName);
+    }
+
+    // TODO could add more tags like http.url
+    return spanBuilder.WithTag(Tags.SpanKind, Tags.SpanKindServer).StartActive(true);
+}
+```
+
+The logic here is similar to the client side instrumentation, except that we are using `_tracer.Extract`
+and tagging the span as `span.kind=server`. We are extracting the headers via `TextMapExtractAdapter`.
+
+Now change the `/api/format/helloString` handler method to use `StartServerSpan`:
+
+```csharp
+[HttpGet("{helloString}", Name = "GetFormat")]
+public string Get(string helloString)
```

Comment:
this is confusing. The operations we have are `sayHello`, `formatString`, and `printHello`. Shouldn't this function be `public string FormatString(string helloTo)`?

Reply:
In ASP.NET controllers, it's common to use the HTTP verb as method name if it's the main endpoint of the ressource. So `Get` for the HTTP GET on this format controller (`GET /api/format` and `GET /api/format/helloTo`). But you are right, it should be `helloTo` and not `helloString`.

Reply:
> it's common to use the HTTP verb as method name

seems redundant given the HttpGet annotation. What if you have more than one endpoint handled in the same class?

Reply:
By default, URLs are parsed in three parts by ASP.NET. First after API is the controller, saying which controller class to use. Second is the action, which usually maps easiest to the method name. Last one is Input Parameter / ID. The HttpGet verb just limits what verbs are allowed. If not given, also i.e. POST would work. If no action is given, the verb or `Index` are used by default.

So in the `Get` case, it wouldn’t be needed to add the attribute. But I think it’s good to have it everywhere to keep it clean. May look strange is such a small example.